pmd-plsql/etc/grammar/PldocAST.jjt

```diff
@@ -3166,36 +3168,27 @@ ASTUnaryExpressionNotPlusMinus UnaryExpressionNotPlusMinus() #UnaryExpressionNot
 {
   (
     <NOT> {sb.append(" NOT "); } simpleNode = UnaryExpression(false)
+  |
+    LOOKAHEAD(2) simpleNode = CaseExpression()
```

Comment:
Do we need the case expression here? I would be part of the PrimaryExpression/PrimaryPrefix anyway... Adding it here would change the AST...

Reply:
All those changes are needed to remove IsOfTypeCondition() from inside IsNullCondition

```ASTIsNullCondition IsNullCondition() #IsNullCondition(>1) :  //yanzin
{ PLSQLNode simpleNode = null; PLSQLNode name = null; StringBuilder sb = new StringBuilder(); }
{
  (
    LOOKAHEAD(PrimaryExpression() <IS> [<NOT>] <NULL>)
    (
        (name = PrimaryExpression() #Name) {sb.append(name.getImage());} <IS> {sb.append(" IS");} [<NOT> {sb.append(" NOT");}] <NULL> {sb.append(" NULL");}
    )
    |
    (
        **simpleNode = IsOfTypeCondition()**
    )
    {
        sb.append(simpleNode.getImage());
    }
  )
  {
    jjtThis.setImage(sb.toString());
	return jjtThis;
  }
}```

Reply:
The case expression problem was actually wrong ordering in PrimaryPrefix, that's why it wasn't parsed (see #1885).

I agree, that having the IsOfTypeCondition nested within IsNullCondition is not nice, but that's the way it is right now. Changing this could potentially break existing rules.

I don't see, that this is a problem when IsNullCondition is used within a Condition (in a WhereClause) - since we already verified with a lookahead that a IsNullCondition is following, so it never parses into IsOfTypeCondition. And for the latter, we have a seperate lookahead within a Condition.